### PR TITLE
optimize EscapeAscii's Display  and CStr's Debug

### PR DESCRIPTION
**Pull request by the8472**
```
old:
    ascii::bench_ascii_escape_display_mixed      17.97µs/iter +/- 204.00ns
    ascii::bench_ascii_escape_display_no_escape 545.00ns/iter   +/- 6.00ns
new:
    ascii::bench_ascii_escape_display_mixed      4.99µs/iter +/- 56.00ns
    ascii::bench_ascii_escape_display_no_escape 91.00ns/iter  +/- 1.00ns
```

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@bors r+

r? @Mark-Simulacrum 

It might be nice to check the code size impact, but my guess is that this isn't really a huge amount of new code. Not sure how much it matters anyway in terms of whether this gets used...

File: library/core/src/iter/adapters/fuse.rs [hunk 1]
```diff
@@ -24,6 +24,10 @@ impl<I> Fuse<I> {
     pub(in crate::iter) fn new(iter: I) -> Fuse<I> {
         Fuse { iter: Some(iter) }
     }
+
+    pub(crate) fn into_inner(self) -> Option<I> {
```
Comment (klensy) [inline library/core/src/iter/adapters/fuse.rs:28]: Inline or it inlines itself good enough?

File: library/core/src/iter/adapters/map.rs [hunk 2]
```diff
@@ -68,6 +68,10 @@ impl<I, F> Map<I, F> {
     pub(in crate::iter) fn new(iter: I, f: F) -> Map<I, F> {
         Map { iter, f }
     }
+
+    pub(crate) fn into_inner(self) -> I {
```
Comment (klensy) [inline library/core/src/iter/adapters/map.rs:72]: here too.

File: library/core/src/iter/adapters/fuse.rs (hunk 1 above)
Comment (the8472) [reply library/core/src/iter/adapters/fuse.rs:28]: Well, it worked in the microbenchmarks at least. With mir inlining, the cross-crate auto-inlining which was recently added and std being compiled with 1CGU that's probably enough.
If perf shows regressions I'll reconsider.

**Pull request merged**